src/einsteinpy/metric/base_metric.py

```diff
@@ -0,0 +1,553 @@
+import warnings
+
+import numpy as np
+
+from einsteinpy import constant
+
+_c = constant.c.value
+_G = constant.G.value
+_Cc = constant.coulombs_const.value
+
+
+class BaseMetric:
+    """
+    Class for defining general Metric Tensors
+
+    To be used in perturbative treatment of EFE and its solutions
+
+    """
+
+    def __init__(
+        self,
+        coords,
+        M,
+        a=0,
+        Q=0,
+        name="Base Metric",
+        metric_cov=None,
+        christoffels=None,
+        f_vec=None,
+        perturbation=None,
+    ):
+        """
+        Constructor
+
+        Parameters
+        ----------
+        coords : string
+            Coordinate system, in which Metric is to be represented
+            "S" - Schwarzschild: Only applicable to Schwarzschild solutions
+            "BL" - Boyer-Lindquist: Applicable to Kerr-Newman solutions
+            "KS" - Kerr-Schild: Useful for adding perturbations to Kerr-Newman solutions
+        M : float
+            Mass of gravitating body, e.g. Black Hole
+        a : float
+            Spin Parameter
+            Defaults to ``0``
+        Q : float
+            Charge on gravitating body, e.g. Black Hole
+            Defaults to ``0``
+        name : str
+            Name of the Metric Tensor. Defaults to ``"Base Metric"``
+        metric_cov : function
+            Function, defining Covariant Metric Tensor
+            It should return a real-valued tensor (2D Array), at supplied coordinates
+            Defaults to ``None``
+            Consult pre-defined classes for function definition
+        christoffels : function
+            Function, defining Christoffel Symbols
+            It should return a real-valued (4,4,4) array, at supplied coordinates
+            Defaults to ``None``
+            Consult pre-defined classes for function definition
+        f_vec : function
+            Function, defining RHS of Geodesic Equation
+            It should return a real-valued (8) vector, at supplied coordinates
+            Defaults to ``None``
+            Consult pre-defined classes for function definition
+        perturbation : function
+            Function, defining Covariant Perturbation Tensor
+            It should return a real-valued tensor (2D Array), at supplied coordinates
+            Defaults to ``None``
+            Function definition similar to ``metric_cov``
+
+        """
+        self.name = name
+        self.coords = coords
+        self.M = M
+        self.a = a
+        self.Q = Q
+        self.metric_cov = metric_cov
+        self.christoffels = christoffels
+        self.f_vec = f_vec
+        self.perturbation = perturbation
```

Comment:
I think, we discussed to add a warning here? I have done so, in `metric_covariant()` and `metric_contravariant()` methods. Implementation per se is sound. Just that, we don't have checks to ensure that the perturbed metric is physical. But, there are cases, where we might want to disregard some physical condition and still do the calculations. This can help with that. It's like a caution-watch-your-step kind of experimental feature, for which the user is warned, and we can expand it with physical checks in the future, to make it more robust.

Reply:
The warning will be up with my next commit.

Reply:
Warning is at 2 places now. @shreyasbapat

Reply:
Great.